diff --git a/libtgvoip.gyp b/libtgvoip.gyp
index 8639848..10cd0e7 100644
--- a/libtgvoip.gyp
+++ b/libtgvoip.gyp
@@ -382,8 +382,12 @@
               'defines': [
                 'WEBRTC_POSIX',
               ],
-              'cflags_cc': [
-                '-msse2',
+              'conditions': [
+                [ '"<!(uname -p)" == "x86_64" or "<!(uname -p)" == "i686"', {
+                  'cflags_cc': [
+                    '-msse2',
+                  ],
+                }]
               ],
               'direct_dependent_settings': {
                 'libraries': [
