diff --git a/Telegram/SourceFiles/logs.cpp b/Telegram/SourceFiles/logs.cpp
index 5ecf2b7..0c7bc64 100644
--- a/Telegram/SourceFiles/logs.cpp
+++ b/Telegram/SourceFiles/logs.cpp
@@ -324,7 +324,11 @@ void start(not_null<Core::Launcher*> launcher) {
 #if defined Q_OS_MAC || defined Q_OS_LINUX
 	} else {
 #ifdef _DEBUG
-		cForceWorkingDir(cExeDir());
+		if (cWorkingDir().isEmpty()) {
+			cForceWorkingDir(cExeDir());
+		} else {
+			cForceWorkingDir(cWorkingDir());
+		}
 #else // _DEBUG
 		if (!cWorkingDir().isEmpty()) {
 			// This value must come only from the "-workdir" argument.
